rpc: add target to getmininginfo field and show next block info

Sjors opened the pull request:
**tl&dr for consensus-code only reviewers**: the first commit splits `CheckProofOfWorkImpl()` in order to create a `DeriveTarget()` helper. The rest of this PR does not touch consensus code. 


There are three ways to represent the proof-of-work in a block:

1. nBits
2. Difficulty
3. Target

The latter notation is useful when you want to compare share work against either the pool target (to get paid) or network difficulty (found an actual block). E.g. for difficulty 1 which corresponds to an nBits value of `0x00ffff`: 

```
share hash: f6b973257df982284715b0c7a20640dad709d22b0b1a58f2f88d35886ea5ac45
target:     7fffff0000000000000000000000000000000000000000000000000000000000
```

It's immediately clear that the share is invalid because the hash is above the target.

This type of logging is mostly done by the pool software. It's a nice extra convenience, but not very important. It impacts the following RPC calls:

1. `getmininginfo` displays the `target` for the tip block
2. `getblock` and `getblockheader` display the `target` for a specific block (ditto for their REST equivalents)

The `getdifficulty` method is a bit useless in its current state, because what miners really want to know if the difficulty for the _next_ block. So I added a boolean argument `next` to `getdifficulty`. (These values are typically the same, except for the first block in a retarget period. On testnet3 / testnet4 they change when no block is found after 20 minutes).

Similarly I added a `next` object to `getmininginfo` which shows `bit`, `difficulty` and `target` for the next block.

In order to test the difficulty transition, an alternate mainnet chain with 2016 blocks was generated and used in `mining_mainnet.py`. The chain is deterministic except for its timestamp and nonce values, which are stored in `mainnet_alt.json`.

As described at the top, this PR introduces a helper method `DeriveTarget()` which is split out from `CheckProofOfWorkImpl`. The proposed `checkblock` RPC in #31564 needs this helper method internally to figure out the consensus target.

Finally, this PR moves `pow.cpp` and `chain.cpp` from `bitcoin_node` to `bitcoin_common`, in order to give `rpc/util.cpp` (which lives in `bitcoin_common`) access to `pow.h`.

Sjors committed:
build: move pow and chain to bitcoin_common

The next commit needs pow.cpp in rpc/util.cpp.

tdb3 reviewed (commented):
Approach ACK

Welcome addition

Sanity checked on mainnet (at block 200,000)
```
$ build/src/bitcoin-cli gettarget
00000000000005db8b0000000000000000000000000000000000000000000000
$ build/src/bitcoin-cli getmininginfo
{
  "blocks": 200000,
  "difficulty": 2864140.507810974,
  "target": "00000000000005db8b0000000000000000000000000000000000000000000000",
  "networkhashps": 26411459642253.2,
  "pooledtx": 0,
  "chain": "main",
  "warnings": [
    "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications"
  ]
}
```

Sjors committed:
rpc: add target to getmininginfo result

Sjors committed:
rpc: add next to getmininginfo

Obtain difficulty and target for the next block without having to call
getblocktemplate.

Sjors commented on test/functional/rpc_blockchain.py:
```diff
@@ -438,6 +439,12 @@ def _test_getdifficulty(self):
+        self.log.info("Test gettarget")
+        target = self.nodes[0].gettarget()
+        # Target corresponding to nBits 0xffff7f20
+        assert_equal(target, "7fffff0000000000000000000000000000000000000000000000000000000000")
```
Done and also added test coverage for nBits.

Sjors commented:
Rebased and split things across more commits.

I also realized that `getmininginfo`, `getdifficulty` and `gettarget` always refer to the tip, rather than the upcoming block which is what miners actually care about. So added a `next` argument to `getdifficulty` and `gettarget` and a `next` object to `getmininginfo`. You'll have to test this manually, because regtest doesn't have difficulty adjustment.

I might push a test that uses the testnet4 genesis block and the rule that difficulty resets to 1 after 20 minutes.

Sjors commented:
Added a helper function `NextEmptyBlockIndex`, which also fixes the missing nBits value on mainnet and signet.

tdb3 commented on src/rpc/mining.cpp:
```diff
@@ -421,11 +422,20 @@ static RPCHelpMan getmininginfo()
                         {RPCResult::Type::NUM, "pooledtx", "The size of the mempool"},
                         {RPCResult::Type::STR, "chain", "current network name (" LIST_CHAIN_NAMES ")"},
                         {RPCResult::Type::STR_HEX, "signet_challenge", /*optional=*/true, "The block challenge (aka. block script), in hexadecimal (only present if the current network is a signet)"},
+                        {RPCResult::Type::OBJ, "next", "The next block, if found now",
+                        {
+                            {RPCResult::Type::NUM, "height", "The next height"},
+                            {RPCResult::Type::STR_HEX, "bits", "The next target nBits"},
+                            {RPCResult::Type::NUM, "difficulty", "The next difficulty"},
+                            {RPCResult::Type::STR_HEX, "target", "The next target"}
+                        }},
```
Let's add release notes for this, `getdifficulty`, and `gettarget`, e.g. something like:

```
Updated RPCs
---
- `getmininginfo` now returns the current target in the `target` field, and a `next` object, which specifies the `height`, `nBits`, `difficulty`, and `target` for the next block.
- `getdifficulty` can now return the difficulty for the next block (rather than the current tip) when calling with the boolean `next` argument set to true.

New RPCs
---
- `gettarget` can be used to return the current target (for tip) or target for the next block (with the `next` argument)
```

tdb3 reviewed (commented):
Approach ACK

Planning to test later today as time allows. Left a minor comment, but it could wait until after testing (to avoid churn).

Sjors committed:
test: check difficulty adjustment using alternate mainnet

Sjors commented:
Added a testnet4 test, which includes the first 2016 real blocks. This takes up 1 MB (IIUC git effectively gzips that into 250kb).

Also added release notes.

I moved the `NextEmptyBlockIndex` helper from `node/miner.h` to `rpc/util.h`.

Sjors commented:
> I moved the `NextEmptyBlockIndex` helper from `node/miner.h` to `rpc/util.h`.

That didn't work, because `rpc/util.cpp` lives in `libbitcoin_common`, while the `UpdateTime` function it calls lives in `libbitcoin_node`. I moved it to  `rpc/server_util.{h,cpp}` instead.

tdb3 commented on test/functional/mining_testnet.py:
```diff
@@ -0,0 +1,64 @@
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.setup_clean_chain = True
+        self.chain = "testnet4"
```
In the future, when testnet4 replaces testnet3 as `-testnet` (no trailing number), we'll need to drop the `4`.

Sjors replied:
Maybe, depends on how soon testnet5 shows up :-)

tdb3 commented on test/functional/mining_testnet.py:
```diff
@@ -0,0 +1,64 @@
+    def add_options(self, parser):
+        parser.add_argument(
+            '--datafile',
+            default='data/testnet4.hex',
```
Similarly here (file rename in the future)

Sjors replied:
Note that this pattern is copied from a similar testnet3 test.

tdb3 reviewed (commented):
Seems like there is some history here regarding regtest and fPowNoRetargeting=true (https://github.com/bitcoin/bitcoin/pull/6853). When configured to avoid internet connectivity, it seems reasonable to me to use testnet4 and canned data (`data/testnet4.hex`).

The bitcoin.conf for the test node has `dnsseed=0`, `fixedseeds=0`, and `connect=0`, which should help avoid internet traffic. Added a sleep in `run_test()` and didn't see internet traffic from the test node in Wireshark.

Planning to circle back and test a bit on mainnet.

Sjors commented on test/functional/rpc_blockchain.py:
```diff
@@ -443,11 +443,13 @@ def _test_getdifficulty(self):
         # 1 hash in 2 should be valid, so difficulty should be 1/2**31
         # binary => decimal => binary math is why we do this check
         assert abs(difficulty * 2**31 - 1) < 0.0001
+        assert_equal(self.nodes[0].getdifficulty(next=True), difficulty)
```
Will do if I need to retouch.

tdb3 reviewed (approved):
ACK 7f34802fd726b259f7df6f605e1d488faf251127


<details><summary>Tested on mainnet for 876960 difficulty adjustment</summary>

Expected: same diff/target in next block
```
$ build/src/bitcoin-cli getblockhash 876959
000000000000000000002428ab473fa8570115d4a000e637814b86248d35d370
$ build/src/bitcoin-cli invalidateblock 000000000000000000002428ab473fa8570115d4a000e637814b86248d35d370
$ build/src/bitcoin-cli getdifficulty
108522647629298.2
$ build/src/bitcoin-cli getdifficulty true
108522647629298.2
$ build/src/bitcoin-cli gettarget
0000000000000000000297fa0000000000000000000000000000000000000000
$ build/src/bitcoin-cli gettarget true
0000000000000000000297fa0000000000000000000000000000000000000000
$ build/src/bitcoin-cli getmininginfo
{
  "blocks": 876958,
  "bits": "170297fa",
  "difficulty": 108522647629298.2,
  "target": "0000000000000000000297fa0000000000000000000000000000000000000000",
  "networkhashps": 8.339496073211752e+20,
  "pooledtx": 3795,
  "chain": "main",
  "next": {
    "height": 876959,
    "bits": "170297fa",
    "difficulty": 108522647629298.2,
    "target": "0000000000000000000297fa0000000000000000000000000000000000000000"
  },
  "warnings": [
    "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications"
  ]
}
$ build/src/bitcoin-cli reconsiderblock 000000000000000000002428ab473fa8570115d4a000e637814b86248d35d370
```

Expected: new diff/target in next block
```
$ build/src/bitcoin-cli getblockhash 876960
0000000000000000000139773a9f91a256b5472f619b4bd297d07836a6587ffe
$ build/src/bitcoin-cli invalidateblock 0000000000000000000139773a9f91a256b5472f619b4bd297d07836a6587ffe
$ build/src/bitcoin-cli getdifficulty
108522647629298.2
$ build/src/bitcoin-cli getdifficulty true
109782075598905.2
$ build/src/bitcoin-cli gettarget
0000000000000000000297fa0000000000000000000000000000000000000000
$ build/src/bitcoin-cli gettarget true
00000000000000000002905c0000000000000000000000000000000000000000
$ build/src/bitcoin-cli getmininginfo
{
  "blocks": 876959,
  "bits": "170297fa",
  "difficulty": 108522647629298.2,
  "target": "0000000000000000000297fa0000000000000000000000000000000000000000",
  "networkhashps": 8.214207064313692e+20,
  "pooledtx": 3914,
  "chain": "main",
  "next": {
    "height": 876960,
    "bits": "1702905c",
    "difficulty": 109782075598905.2,
    "target": "00000000000000000002905c0000000000000000000000000000000000000000"
  },
  "warnings": [
    "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications"
  ]
}
$ build/src/bitcoin-cli reconsiderblock 0000000000000000000139773a9f91a256b5472f619b4bd297d07836a6587ffe
```

Expected: same diff/target in next block
```
$ build/src/bitcoin-cli getblockhash 876961
000000000000000000024565da14cc4336d8713bcadbc889cb22ed73b940e048
$ build/src/bitcoin-cli invalidateblock 000000000000000000024565da14cc4336d8713bcadbc889cb22ed73b940e048
$ build/src/bitcoin-cli getdifficulty
109782075598905.2
$ build/src/bitcoin-cli getdifficulty true
109782075598905.2
$ build/src/bitcoin-cli gettarget
00000000000000000002905c0000000000000000000000000000000000000000
$ build/src/bitcoin-cli gettarget true
00000000000000000002905c0000000000000000000000000000000000000000
$ build/src/bitcoin-cli getmininginfo
{
  "blocks": 876960,
  "bits": "1702905c",
  "difficulty": 109782075598905.2,
  "target": "00000000000000000002905c0000000000000000000000000000000000000000",
  "networkhashps": 8.133301750223401e+20,
  "pooledtx": 4022,
  "chain": "main",
  "next": {
    "height": 876961,
    "bits": "1702905c",
    "difficulty": 109782075598905.2,
    "target": "00000000000000000002905c0000000000000000000000000000000000000000"
  },
  "warnings": [
    "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications"
  ]
}
$ build/src/bitcoin-cli reconsiderblock 000000000000000000024565da14cc4336d8713bcadbc889cb22ed73b940e048
```
</details>

tdb3 committed:
test: use DIFF_1_N_BITS in tool_signet_miner

tdb3 committed:
test: use REGTEST_N_BITS in feature_block

tdb3 commented:
Looks like we can re-use the new `DIFF_1_N_BITS` and `REGTEST_N_BITS` in a few other tests (https://github.com/tdb3/bitcoin/commits/2024/12/gettarget/). Could be added if there are other changes or left for future follow-up. The use of `REGTEST_N_BITS` might be borderline (the test is creating blocks by hand).

Sjors commented:
@tdb3 thanks, I'll cherry-pick those two commits if any other changes are needed here.

ryanofsky commented on test/functional/test_framework/blocktools.py:
```diff
@@ -65,6 +65,10 @@
+REGTEST_N_BITS = 0x207fffff  # difficulty retargeting is disabled in REGTEST chainparams"
+
+def nbits_str(nbits):
+    return nbits.to_bytes(4, "big").hex()
```
In commit "test: add nBits coverage" (ecb18b3405e4a61f872d03f667c167fb7c4a52cf)

`return f"{nbits:08x}"` would be a more direct conversion from number to string and would make the python code more consistent with the c++ code.

Sjors replied:
Taken

ryanofsky reviewed (commented):
Code review 7f34802fd726b259f7df6f605e1d488faf251127. I got pretty lost here and only reviewed about half of it so far, but overall I think the changes look good and make a lot of sense.

I left some suggestions below which I think could make code clearer, but they are not important so feel free to ignore any you disagree with.

Sjors committed:
consensus: add DeriveTarget() to pow.h

Split CheckProofOfWorkImpl() to introduce a helper function
DeriveTarget() which converts the nBits value to the target.

The function takes pow_limit as an argument so later commits can
avoid having to pass ChainstateManager through the call stack.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

Sjors committed:
Add target to getblock(header) in RPC and REST

Sjors committed:
doc: add release notes

Co-Authored-By: tdb3 <106488469+tdb3@users.noreply.github.com>

Sjors commented on test/functional/test_framework/blocktools.py:
```diff
@@ -66,10 +67,15 @@
     return nbits.to_bytes(4, "big").hex()
 
+def target_str(target):
+    return ser_uint256(target)[::-1].hex()
```
Taken

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -553,7 +553,7 @@ static RPCHelpMan getblockheader()
                             {RPCResult::Type::NUM_TIME, "mediantime", "The median block time expressed in " + UNIX_EPOCH_TIME},
                             {RPCResult::Type::NUM, "nonce", "The nonce"},
-                            {RPCResult::Type::STR_HEX, "bits", "The bits"},
+                            {RPCResult::Type::STR_HEX, "bits", "The target nBits"},
```
Ok, I used "nBits: compact representation of the block difficulty target".

To avoid confusion, I introduced a new commit 45e2edde903e71fbb085860af781f5ac65095579 that adds a "target" field to `getblockheader` and `getblock` (as well as their REST equivalents).

Sjors commented on src/pow.cpp:
```diff
@@ -11,6 +11,20 @@
 #include <uint256.h>
 #include <util/check.h>
 
+bool DeriveTarget(unsigned int nBits, const uint256 pow_limit, arith_uint256& target)
```
Done in 6accee18442f79fd2f2796027371a70e3757d825.

Sjors commented:
Thanks for the review @ryanofsky. I took your suggestion to split `CheckProofOfWorkImpl`. Can you add the Consensus label to this PR?

@tdb3 I took your two patches and one suggestion

tdb3 reviewed (commented):
Approach ACK

Reviewed code changes.

Syncing from genesis on mainnet (because we're touching `CheckProofOfWorkImpl()`, I don't mind being paranoid, and it happens to be the genesis block anniversary today), so will report back (with ACK if all is well).

Looks like we could add `target` to `getblockchaininfo` and `getchainstates` as well (the top two commits on https://github.com/tdb3/bitcoin/commits/2024/12/gettarget/ add target and associated tests). Could be left for follow up if desired.

ryanofsky commented on test/functional/mining_testnet.py:
```diff
@@ -0,0 +1,64 @@
+    def add_options(self, parser):
+        parser.add_argument(
+            '--datafile',
+            default='data/testnet4.hex',
```
In commit "test: check testnet4 difficulty adjustment" (6b504d6cf8d0b3cf14b0df1683d080f505588e6b)

Can there be some comment about how the testnet.4 file is generated and what it contains? Maybe there could be a readme in the test/functional/data/ directory.

It seems ok to have one test like this with a large static data file, but it would be nice in the future to be able to generate block data more dynamically, and having a description of the data could help implement that.

Sjors replied:
Just this embarrassing bash script:

```sh
for i in {0..2015}
do                                 
hash=`build/src/bitcoin-cli -testnet4 getblockhash $i`
block=`build/src/bitcoin-cli -testnet4 getblock $hash 0`
echo $block >> test/functional/data/testnet4.hex
done
```

ryanofsky reviewed (approved):
Code review ACK b35ed41b03578586e380cb73aece14046ec2da93. Nice changes that make the mining interface more useful and complete, and are cleanly implemented

tdb3 reviewed (approved):
Code review re ACK b35ed41b03578586e380cb73aece14046ec2da93

Sync from genesis was successful (to 877742, `000000000000000000008bac930b7efd54a7ddef0b9ec5b1c0c00b6a6c8e0cfb`)

fyi, conf used:
```
[main]
connect=<sister node on LAN>
checkpoints=0
prune=53000
dbcache=24000
```

Sjors committed:
rpc: add target and bits to getblockchaininfo

Sjors committed:
rpc: add target and bits to getchainstates

Sjors commented:
Small changes:
* dropped testnet4 genesis block
* explained in the test how to generate / verify testnet4.hex
* added `bits` and `target` to `getblockchaininfo` and `getchainstates` (didn't realise @tdb3 had already implemented it)

tdb3 commented on src/rpc/blockchain.cpp:
```diff
@@ -1395,12 +1397,16 @@ RPCHelpMan getblockchaininfo()
+    obj.pushKV("bits", strprintf("%08x", tip.nBits));
+    obj.pushKV("target", GetTarget(tip, pow_limit).GetHex());
     obj.pushKV("difficulty", GetDifficulty(tip));
+    obj.pushKV("target", GetTarget(tip, pow_limit).GetHex());
```
Looks like this line is a duplicate (`target` added twice)

Sjors replied:
Fixed

tdb3 reviewed (commented):
Approach ACK

Verified the addition of `bits` and `target` to `getblockchaininfo` and `getchainstates` on mainnet:
```
$ build/src/bitcoin-cli getblockchaininfo
{
  "chain": "main",
  "blocks": 877759,
  "headers": 877796,
  "bestblockhash": "00000000000000000001cdf88323d957f92d07693b953c32646bac7c797634af",
  "bits": "1702905c",
  "target": "00000000000000000002905c0000000000000000000000000000000000000000",
  "difficulty": 109782075598905.2,
  "time": 1735981113,
  "mediantime": 1735976878,
  "verificationprogress": 0.9998884052581367,
  "initialblockdownload": false,
  "chainwork": "0000000000000000000000000000000000000000a50ce26166c7c567206d3760",
  "size_on_disk": 55429575203,
  "pruned": true,
  "pruneheight": 848361,
  "automatic_pruning": true,
  "prune_target_size": 55574528000,
  "warnings": [
    "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications"
  ]
}
$ build/src/bitcoin-cli getchainstates
{
  "headers": 877796,
  "chainstates": [
    {
      "blocks": 877778,
      "bestblockhash": "0000000000000000000036bc3aba12ed13cc759cba2a55309d3b3cd3d9183a8a",
      "bits": "1702905c",
      "target": "00000000000000000002905c0000000000000000000000000000000000000000",
      "difficulty": 109782075598905.2,
      "verificationprogress": 0.999934824273136,
      "coins_db_cache_bytes": 8388608,
      "coins_tip_cache_bytes": 24081596416,
      "validated": true
    }
  ]
}
```

tdb3 reviewed (approved):
Code review re ACK 223e8bcb9990834a5c9ffcd8a79f617a6f2ecfc1

Sjors commented on doc/release-notes-31583.md:
```diff
@@ -0,0 +1,14 @@
+
+REST interface
+---
+- `GET /rest/block/<BLOCK-HASH>` and `GET /rest/headers/<BLOCK-HASH>` now return the current target in the `target` field
```
For the web developer in me that's obvious :-)

But I'll add `.json` to make it more clear this doesn't apply to the other format.

Sjors commented on test/functional/mining_testnet.py:
```diff
@@ -0,0 +1,73 @@
+do
+  hash=`bitcoin-cli -testnet4 getblockhash $i`
+  block=`bitcoin-cli -testnet4 getblock $hash 0`
+  echo $block >> data/testnet4.hex
```
I thought about generating a fake testnet4 with empty blocks, but that wouldn't be much smaller. However if testnet5 turns out to have large initial blocks, we can always do that.

Headers are not enough, because I need the tip to update. Unless you want to revive #9483 :-)

Sjors replied:
That said, committing only 2016 nonces would make it a lot smaller, so maybe it's a good idea regardless.

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -442,7 +445,58 @@ static RPCHelpMan getdifficulty()
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
+    ChainstateManager& chainman = EnsureAnyChainman(request.context);
+    LOCK(cs_main);
+    CBlockIndex* tip{CHECK_NONFATAL(chainman.ActiveChain().Tip())};
```
I had to move `CHECK_NONFATAL` outside the `WITH_LOCK` call to avoid `returning reference to local temporary object [-Wreturn-stack-address]`

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -781,6 +843,10 @@ static RPCHelpMan getblock()
             throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
         }
+
+        // Now that we have a lock on cs_main, get powLimit so blockToJSON()
+        // doesn't have to.
```
I switched to using `chainman.GetConsensus()` everywhere.

The lock on cs_main is needed to get `pblockindex` and `tip`, but it's not needed to get `chainman` and therefore also not needed to get `pow_limit`. I'll move some things around.

Sjors commented on src/rpc/server_util.cpp:
```diff
@@ -129,3 +133,18 @@ AddrMan& EnsureAnyAddrman(const std::any& context)
     return EnsureAddrman(EnsureAnyNodeContext(context));
 }
+
+void NextEmptyBlockIndex(CBlockIndex* tip, const Consensus::Params& consensusParams, CBlockIndex& next_index)
+{
+    CBlockHeader next_header{};
+    next_header.hashPrevBlock  = tip->GetBlockHash();
```
This results in `assigning to 'CBlockIndex *' from 'const CBlockIndex *' discards qualifier` at the line `next_index.pprev = &tip;`.

I'd rather not enter the rabbit hole of changing `CBlockIndex` to make `pprev` as `const CBlockIndex*`. So for now I just drop the `const` qualifier for `tip`. Unless you have another suggestion?

Sjors committed:
rpc: add GetTarget helper

Sjors committed:
rpc: add nBits to getmininginfo

Also expands nBits test coverage.

Sjors commented:
Addressed comments. Looking into replacing `testnet4.hex` with something much more compact.

Sjors committed:
Use OP_0 for BIP34 padding in signet and tests

For blocks 1 through 15 the script_BIP34_coinbase_height appends OP_1
to comply with BIP34 and avoid bad-cb-length.

This is inconsistent with BlockAssembler::CreateNewBlock() which adds
OP_0 instead.

The utxo_total_supply fuzzer and MinerTestingSetup::Block also use OP_0.

Changing it is required to import the test vectors in the next commit.

It also ensures the test vectors can be regenerated using the CPU miner
at https://github.com/pooler/cpuminer without patches (it uses OP_0).

The same helper is used by the signet miner, so this will impact newly
bootstrapped signets.

ryanofsky reviewed (approved):
Code review ACK d925a9bd7a04f068ede227addbaef3025b9793b9. Lot of nice simplifications since last review: using references instead of pointers, getting rid of intermediate variables, using chainman.GetConsensus(), also improving release notes and test documentation and updating more RPCs to return the bits and target fields.

It would be really nice to remove the large test data file, so happy to rereview when the new test is ready.

tdb3 reviewed (approved):
Code review re ACK d925a9bd7a04f068ede227addbaef3025b9793b9

Great updates.
Would review again if an alternate test approach for `data/testnet4.hex` was used. I also like the idea of possibly using early mainnet (wouldn't need to be changed). Weighing in at ~903KB (hex), using mainnet is smaller than the 1.2MB of testnet4 block data. A nonce approach could be even nicer/smaller.

There are other areas where `.GetParams().GetConsensus()` could be replaced with `.GetConsensus()`, but I think it's prudent to prevent touching other code unnecessarily and to prevent further scope creep on this PR.

<details>
<summary>Diff of other `GetConsensus()` areas (for posterity)</summary>

```diff
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index 1070e8f68f..abc4b44b44 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -1244,7 +1244,7 @@ static RPCHelpMan verifychain()
 
     Chainstate& active_chainstate = chainman.ActiveChainstate();
     return CVerifyDB(chainman.GetNotifications()).VerifyDB(
-               active_chainstate, chainman.GetParams().GetConsensus(), active_chainstate.CoinsTip(), check_level, check_depth) == VerifyDBResult::SUCCESS;
+               active_chainstate, chainman.GetConsensus(), active_chainstate.CoinsTip(), check_level, check_depth) == VerifyDBResult::SUCCESS;
 },
     };
 }
@@ -1410,7 +1410,7 @@ RPCHelpMan getblockchaininfo()
     }
     if (chainman.GetParams().GetChainType() == ChainType::SIGNET) {
         const std::vector<uint8_t>& signet_challenge =
-            chainman.GetParams().GetConsensus().signet_challenge;
+            chainman.GetConsensus().signet_challenge;
         obj.pushKV("signet_challenge", HexStr(signet_challenge));
     }
 
@@ -1781,7 +1781,7 @@ static RPCHelpMan getchaintxstats()
 {
     ChainstateManager& chainman = EnsureAnyChainman(request.context);
     const CBlockIndex* pindex;
-    int blockcount = 30 * 24 * 60 * 60 / chainman.GetParams().GetConsensus().nPowTargetSpacing; // By default: 1 month
+    int blockcount = 30 * 24 * 60 * 60 / chainman.GetConsensus().nPowTargetSpacing; // By default: 1 month
 
     if (request.params[1].isNull()) {
         LOCK(cs_main);
@@ -2121,7 +2121,7 @@ static RPCHelpMan getblockstats()
     ret_all.pushKV("minfeerate", (minfeerate == MAX_MONEY) ? 0 : minfeerate);
     ret_all.pushKV("mintxsize", mintxsize == MAX_BLOCK_SERIALIZED_SIZE ? 0 : mintxsize);
     ret_all.pushKV("outs", outputs);
-    ret_all.pushKV("subsidy", GetBlockSubsidy(pindex.nHeight, chainman.GetParams().GetConsensus()));
+    ret_all.pushKV("subsidy", GetBlockSubsidy(pindex.nHeight, chainman.GetConsensus()));
     ret_all.pushKV("swtotal_size", swtotal_size);
     ret_all.pushKV("swtotal_weight", swtotal_weight);
     ret_all.pushKV("swtxs", swtxs);
diff --git a/src/rpc/mining.cpp b/src/rpc/mining.cpp
index e15c4e62c6..bb193f1ea1 100644
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -817,7 +817,7 @@ static RPCHelpMan getblocktemplate()
         // TODO: Maybe recheck connections/IBD and (if something wrong) send an expires-immediately template to stop miners?
     }
 
-    const Consensus::Params& consensusParams = chainman.GetParams().GetConsensus();
+    const Consensus::Params& consensusParams = chainman.GetConsensus();
 
     // GBT must be called with 'signet' set in the rules for signet chains
     if (consensusParams.signet_blocks && setClientRules.count("signet") != 1) {
diff --git a/src/test/fuzz/p2p_headers_presync.cpp b/src/test/fuzz/p2p_headers_presync.cpp
index 873eb2b1cc..7fb2381fa7 100644
--- a/src/test/fuzz/p2p_headers_presync.cpp
+++ b/src/test/fuzz/p2p_headers_presync.cpp
@@ -135,7 +135,7 @@ CBlock ConsumeBlock(FuzzedDataProvider& fuzzed_data_provider, const uint256& pre
 
 void FinalizeHeader(CBlockHeader& header, const ChainstateManager& chainman)
 {
-    while (!CheckProofOfWork(header.GetHash(), header.nBits, chainman.GetParams().GetConsensus())) {
+    while (!CheckProofOfWork(header.GetHash(), header.nBits, chainman.GetConsensus())) {
         ++(header.nNonce);
     }
 }

```
</details>

Sjors commented:
Here we go!

I generated an alternate mainnet with 2016 blocks, storing timestamps and nonces. The process is explained at the top of the renamed test file: ee1dcc6074323ae361e28cf235ddfc9bf07be1da. 

I did leave out some minutia for readability sake, and because things will change over time. E.g. I had to disable the `miner.isTestChain())` guard in `getblocktemplate`, I patched the CPU miner to work with taproot payout addresses https://github.com/pooler/cpuminer/pull/293 and set the coinbase version to https://github.com/pooler/cpuminer/pull/294. However it's trivial to work around those when reconstructing or expanding these test vectors.

47bf1086dacfc504699d672dc3faf8a77eceee67 changes the BIP34 padding from `OP_1` to `OP_0` in the test suite, but this will also impact new signets (but it shouldn't matter).

tdb3 commented on test/functional/mining_mainnet.py:
```diff
@@ -0,0 +1,121 @@
+are stored in data/mainnet_alt.json and used to reconstruct the chain without
+having to redo the proof-of-work.
+
+The timestamp was not kept constant because at difficulty 1 it's not sufficient
+to only grind the nonce. Grinding the extra_nonce or version field instead
+would have required additional (stratum) software. It would also make it more
+complicated to reconstruct the blocks in this test.
```
Wonder if it would be sufficient to grind only the nonce if starting with a testnet genesis block (instead of mainnet), but not sure it's worth digging into. It might allow removing the timestamps from `mainnet_alt.json` (reducing 35KB of the current 71KB), but the current approach is already more manageable that the previous ~1MB.

Sjors replied:
Testnet3, testnet4 and mainnet all have a difficulty 1 genesis block, so it wouldn't make a difference.

tdb3 reviewed (approved):
code review re ACK 58b828f7f7ad216c7d7bf2e2ff431a66591e9d5c

The data file is down to 71K, which is much nicer!

`mining_mainnet.py` ran locally without issue.

Sjors commented on test/functional/tool_signet_miner.py:
```diff
@@ -55,7 +56,7 @@ def run_test(self):
                 f'--address={node.getnewaddress()}',
                 f'--grind-cmd={self.options.bitcoinutil} grind',
-                '--nbits=1d00ffff',
+                f'--nbits={hex(DIFF_1_N_BITS)}',
```
Fixed

Sjors commented on test/functional/test_framework/blocktools.py:
```diff
@@ -129,7 +129,7 @@ def script_BIP34_coinbase_height(height):
     if height <= 16:
         res = CScriptOp.encode_op_n(height)
         # Append dummy to increase scriptSig size above 2 (see bad-cb-length consensus rule)
```
I changed it to "to 2"

Sjors commented:
I just realized the coinbase transactions are unspendable due to the `unexpected-witness` and hardcode SegWit activation height on mainnet.

It doesn't matter for this PR, but I'm working on a new edition that uses legacy addresses and also checks that the coinbase is spendable. Will push here or, if it's already merged, to a followup.

tdb3 reviewed (approved):
code review re ACK b945668216abbc978ce351b8a89910e651c6e595

Changes were to address the latest comments (removing "0x" and adjusting a comment).

```diff
$ git range-diff 228aba2c4d9ac0b2ca3edd3c2cdf0a92e55f669b..58b828f7f7ad216c7d7bf2e2ff431a66591e9d5c e8ea483fec05d824d2f35182b1a3bd68416b2df3~..b945668216abbc978ce351b8a89910e651c6e595
 1:  6accee1844 =  1:  e8ea483fec consensus: add DeriveTarget() to pow.h
 2:  3e976d4171 =  2:  6a72497f50 build: move pow and chain to bitcoin_common
 3:  2401931542 =  3:  c370fa33c2 rpc: add nBits to getmininginfo
 4:  ddd2ede17a =  4:  c50ea52534 test: use REGTEST_N_BITS in feature_block
 5:  ef9e86039a =  5:  4db86d8638 rpc: gettarget
 6:  a0145ce476 =  6:  0c3b0af556 Add target to getblock(header) in RPC and REST
 7:  f06f55090e =  7:  684e3247d2 rpc: add target to getmininginfo result
 8:  1502b98b8b =  8:  333a146882 rpc: add next to getdifficulty and gettarget
 9:  590d3ab92c =  9:  b0f82495a9 rpc: add target and bits to getblockchaininfo
10:  4463b0e04c = 10:  02ba73a4a0 rpc: add target and bits to getchainstates
11:  a9ce7d1774 = 11:  e9f5b618dc rpc: add next to getmininginfo
12:  47bf1086da ! 12:  047b471a87 Use OP_0 for BIP34 padding in signet and tests
    @@ test/functional/test_framework/blocktools.py: from .script import (
          OP_RETURN,
          OP_TRUE,
      )
    -@@ test/functional/test_framework/blocktools.py: def script_BIP34_coinbase_height(height):
    +@@ test/functional/test_framework/blocktools.py: def add_witness_commitment(block, nonce=0):
    + def script_BIP34_coinbase_height(height):
          if height <= 16:
              res = CScriptOp.encode_op_n(height)
    -         # Append dummy to increase scriptSig size above 2 (see bad-cb-length consensus rule)
    +-        # Append dummy to increase scriptSig size above 2 (see bad-cb-length consensus rule)
     -        return CScript([res, OP_1])
    ++        # Append dummy to increase scriptSig size to 2 (see bad-cb-length consensus rule)
     +        return CScript([res, OP_0])
          return CScript([CScriptNum(height)])
      
13:  ee1dcc6074 = 13:  4138351f60 test: check difficulty adjustment using alternate mainnet
14:  9c646bf4f5 ! 14:  289e304ac3 test: use DIFF_1_N_BITS in tool_signet_miner
    @@ test/functional/tool_signet_miner.py: class SignetMinerTest(BitcoinTestFramework
                      f'--address={node.getnewaddress()}',
                      f'--grind-cmd={self.options.bitcoinutil} grind',
     -                '--nbits=1d00ffff',
    -+                f'--nbits={hex(DIFF_1_N_BITS)}',
    ++                f'--nbits={DIFF_1_N_BITS:08x}',
                      f'--set-block-time={int(time.time())}',
                      '--poolnum=99',
                  ], check=True, stderr=subprocess.STDOUT)
15:  58b828f7f7 = 15:  b945668216 doc: add release notes
```

Good catch about the unspendable coinbase outputs. I'd re-review the updated approach in either case (in this PR or a new one).

Sjors commented on test/functional/mining_mainnet.py:
```diff
@@ -0,0 +1,168 @@
+        block.nTime = blocks['timestamps'][height - 1]
+        block.nBits = DIFF_1_N_BITS
+        block.nNonce = blocks['nonces'][height - 1]
+        block.vtx = copy.deepcopy(txs)
```
dec5f8ba8ff579e3e708079dde8a7ff732884479: @maflcko I'm puzzled by the need to `deepcopy` here, but if I just do `block.vtx = txs` then in the next call to `mine`, without passing any transactions in, `txs` will have the coinbase of the previous block in it.

Sjors replied:
Well the linter forced me to change the `txs` default to `None`, which "solved" the issue.

ismaelsadeeq commented on src/rpc/blockchain.cpp:
```diff
@@ -431,7 +432,9 @@ static RPCHelpMan getdifficulty()
                 "\nReturns the proof-of-work difficulty as a multiple of the minimum difficulty.\n",
-                {},
+                {
+                    {"next", RPCArg::Type::BOOL, RPCArg::Default{false}, "difficulty for the next block, if found now"},
```
@Sjors what do you mean by "if found now" here and the other places?

Sjors replied:
In testnet3 and testnet4 the difficulty depends on the timestamp. After 20 minutes it drops to 1. These methods derive difficulty using the current timestamp.

ismaelsadeeq reviewed (commented):
Concept ACK  
I think the added feature here is useful to miners!

The consensus changes involve splitting `CheckProofOfWorkImpl` into two functions:

1. `DeriveTarget`, which returns the proof-of-work target given a compact target.  
2. `CheckProofOfWorkImpl`, which performs its normal check to verify whether a block hash satisfies the proof-of-work requirements.

This enables the usage of `DeriveTarget` in other places.

The large diff is due to an alternate mainnet chain data used in test 054bbceded4515a2cef7f247aca576f0c4413205.

ryanofsky commented on test/functional/test_framework/blocktools.py:
```diff
@@ -128,8 +128,8 @@ def add_witness_commitment(block, nonce=0):
     if height <= 16:
         res = CScriptOp.encode_op_n(height)
-        # Append dummy to increase scriptSig size above 2 (see bad-cb-length consensus rule)
-        return CScript([res, OP_1])
```
In commit "Use OP_0 for BIP34 padding in signet and tests" (75342f7e19cbbae083c2a3561f2c2cefd4e2968a)

Do we know any reason why this code was using OP_1 instead of OP_0 previously? Any possible advantages to using OP_1? It looks like this code was introduced in #16363

Sjors replied:
I don't know. @maflcko?

ryanofsky commented on test/functional/mining_mainnet.py:
```diff
@@ -0,0 +1,167 @@
+
+        assert_equal(node.getblockcount(), 2014)
+
+        # For the last block of the retarget period, check that previously generated
+        # coins are spendable.
```
In commit "test: check difficulty adjustment using alternate mainnet" (a70f967bda5a0998a2b5a851cd32d969b021df15)

This seems ok, but I think I don't understand why it is useful to test that coins are spendable. It seems like something besides the point of the test (at least as stated in the description), which is to test difficulty retargeting functions on the main chain that can't be tested on the regtest chain. I wonder if this is spending test is checking something that existing tests are not already checking for?

More generally, I think it would be good to update description at top of this file to explain what things this test is trying to check for and cover. IMO, it is hard to parse right now because there is so much detail about how the test data file is generated and not as much information about the test itself. I think it would be better to move detailed information about how the data file is generated to a README file in the data directory, and only include more essential information and a link here.

Sjors replied:
> This seems ok, but I think I don't understand why it is useful to test that coins are spendable.

It's basically checking against a bug I introduced in an earlier version. Initially I used taproot addresses for the coinbase which are unspendable due to the `unexpected-witness".

However it might be simpler if I drop it and just leave a separate commit as a comment in the PR, if someone needs it later.

ryanofsky reviewed (approved):
Code review ACK 712e3ffde5fa4fd26c0f217942b4a289cf37f361. Main change since last review is mining_testnet.py test being replaced by mining_mainnet.py test with a much smaller data file. It was also rebased and there were small documentation changes and string format tweak in mining_mainnet.py

tdb3 reviewed (approved):
code review re ACK 712e3ffde5fa4fd26c0f217942b4a289cf37f361

Since last review, the test was updated to use spendable coinbase UTXOs.

Sjors commented:
I dropped the test code that proves the coinbase outputs are spendable. For posterity, it's in https://github.com/Sjors/bitcoin/commit/ac516671b1846dbc40bea23c59399dbd7ec9b76b.

Added `test/functional/data/README.md` to explain how `mainnet_alt.json` was generated.

ryanofsky commented on test/functional/mining_mainnet.py:
```diff
@@ -0,0 +1,110 @@
+
+# Derived from first BIP32 test vector master key:
+# Use pkh() because tr() outputs at low heights are not spendable (unexpected-witness)
+COINBASE_OUTPUT_DESCRIPTOR="pkh(xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi/44h/0h/0h/<0;1>/*)#fkjtr0yn"
+COINBASE_SCRIPT_PUBKEY="76a914eadbac7f36c37e39361168b7aaee3cb24a25312d88ac"
```
In commit "test: check difficulty adjustment using alternate mainnet" (90b6c189d9b18203e119a5f3478f7b0da7d71392)

Note: this seems ok, but I guess it could be simplified a little since COINBASE_OUTPUT_DESCRIPTOR is unused, and no spending is done anymore.

In case you do want to bring back the spending test that seems fine if it's explained a little. I wasn't necessarily asking to remove it before, but was just confused by why it was there and a little overwhelmed with the details chain generation in the test description. Keeping this is as or simplifying more also both seem fine to me.

Sjors replied:
I'm fine with leaving it out.

I also dropped the `COINBASE_OUTPUT_DESCRIPTOR` constant and moved the descriptor itself to the README.

ryanofsky reviewed (approved):
Code review ACK 39eabfa433e0bb9d85c67c6f9ce74c1570a21417. Since last review just simplified the test and clarified documentation a bit (thanks!)

ryanofsky reviewed (approved):
Code review ACK cb305da72df501991a6700cd20be79dde4591184, since last review moving coinbase key information from the test to the data readme file. I think this change makes both easier to understand now.

tdb3 reviewed (approved):
Code review re ACK cb305da72df501991a6700cd20be79dde4591184

Nice cleanup. It seems reasonable to have removed the spend test.

`git range-diff e7c479495509c068215b73f6df070af2d406ae15..712e3ffde5fa4fd26c0f217942b4a289cf37f361 e7c479495509c068215b73f6df070af2d406ae15..cb305da72df501991a6700cd20be79dde4591184`

ismaelsadeeq commented on test/functional/data/README.md:
```diff
@@ -0,0 +1,42 @@
+## mainnet_alt.json
+
+For easier testing the difficulty is maximally increased in the first (and only)
+regarget period, by producing blocks approximately 2 minutes apart.
```
nit: 
```suggestion
retarget period, by producing blocks approximately 2 minutes apart.
```

Sjors replied:
Will fix typo if I need to retouch.

ismaelsadeeq commented on test/functional/data/README.md:
```diff
@@ -0,0 +1,42 @@
+# Various test vectors
```
Nice readme, I think other data files should also be explained like this!

ismaelsadeeq commented on test/functional/mining_mainnet.py:
```diff
@@ -0,0 +1,108 @@
+
+        assert_equal(node.getblockcount(), 2015)
+
+        self.log.info("Check difficulty adjustment with getmininginfo")
```
Should we also the difficulties before mining those blocks?

Sjors replied:
I'm not sure what you mean here.

Sjors replied:
Do you mean checking that the difficulty is 1 after the genesis block? That seems overkill, since we already check the last block of the first retarget period (2015).

ismaelsadeeq reviewed (commented):
Code review ACK cb305da72df501991a6700cd20be79dde4591184

> The gettarget and getdifficulty methods are a bit useless in their current state, because what miners really want to know if the target / difficulty for the next block. So I added a boolean argument next to getdifficulty and gettarget. (These values are typically the same, except for the first block in a retarget period.

> Similarly I added a next object to getmininginfo which shows bit, difficulty and target for the next block.

Is there a reason we are introducing this redundancy of being able retrieve the same data in two ways (`gettarget` and `getmininginfo`)?

Sjors commented:
> Is there a reason we are introducing this redundancy of being able retrieve the same data in two ways (gettarget and getmininginfo)?

The redundancy was already there with `getdifficulty`. That said, I don't personally need `gettarget` so I could drop it if people prefer.

fanquake commented:
> The redundancy was already there with getdifficulty. That said, I don't personally need gettarget so I could drop it if people prefer.

Some redundancy already existing isn't a reason to make things worse? We don't usually add new RPCs for fun.

Sjors commented:
I'll drop the `gettarget` RPC by modifying 98e8c8fafc96e856bb95bb6e68e54d054e73128d to only introduce the `GetTarget` helper.

Sjors commented:
Along similar lines I also dropped the change that added `next` to `getdifficulty`; miners can use `getmininginfo`. For that I moved the `NextEmptyBlockIndex` helper from f408fdef874b6aa5d9fcb6968a13e5ebdf57b01a to 7921c0b1e82427b8129b21931cb83f45222820e4 -> cf0a62878be214cd4ec779aab214221b27b769b6.

We can probably deprecate `getdifficulty` since it seems completely useless, but I won't do that here.

ismaelsadeeq commented on src/rpc/server_util.cpp:
```diff
@@ -129,3 +133,18 @@ AddrMan& EnsureAnyAddrman(const std::any& context)
     return EnsureAddrman(EnsureAnyNodeContext(context));
 }
+
+void NextEmptyBlockIndex(CBlockIndex& tip, const Consensus::Params& consensusParams, CBlockIndex& next_index)
```
In "rpc: add next to getmininginfo" cf0a62878be214cd4ec779aab214221b27b769b6

You don't need to update this, just thinking out loud: would it be better for this to simply be `NextBlockHeader` and to modify `GetDifficulty` and `GetTarget` to accept the block header since the info they need are already in the header.

Sjors replied:
A `CHeader` does not contain `nHeight`, though the height is trivial anyway.

I'm a bit reluctant to expand the scope of this PR even further by refactoring `GetDifficulty`. So I prefer to leave this as a potential followup.

ismaelsadeeq commented on test/functional/mining_basic.py:
```diff
@@ -213,6 +213,12 @@ def assert_submitblock(block, result_str_1, result_str_2=None):
+            'height': 201,
+            'target': target_str(REGTEST_TARGET),
+            'bits': nbits_str(REGTEST_N_BITS),
+            'difficulty': Decimal('4.656542373906925E-10')
```
nit:
Could be better to define this value as a const just like you did for nBits and target.

Sjors replied:
Will do if I need to retouch.

ismaelsadeeq reviewed (approved):
re-ACK a4df12323c4e9230bca58562ba17ecee4233f8fe

The changes in the last push are:  
- https://github.com/bitcoin/bitcoin/pull/31583#issuecomment-2606863459
- https://github.com/bitcoin/bitcoin/pull/31583#issuecomment-2607000049)
- The addition of `nbits` and `target` to `getchainstates`  RPC result in commit `2d18a078a2d9eaa53b8b7acc7600141c69f0d742`.  


I've Also tested this on my mainnet node.

<details>
<summary>results</summary>

```terminal
bitcoin (2024/12/gettarget)# build/src/bitcoin-cli getmininginfo
{
  "blocks": 880362,
  "bits": "17028c61",
  "difficulty": 110451907374649.5,
  "target": "000000000000000000028c610000000000000000000000000000000000000000",
  "networkhashps": 7.683954895255723e+20,
  "pooledtx": 5044,
  "chain": "main",
  "next": {
    "height": 880363,
    "bits": "17028c61",
    "difficulty": 110451907374649.5,
    "target": "000000000000000000028c610000000000000000000000000000000000000000"
  },
  "warnings": [
    "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications"
  ]
}
bitcoin (2024/12/gettarget)# build/src/bitcoin-cli getmininginfo
{
  "blocks": 880363,
  "bits": "17028c61",
  "difficulty": 110451907374649.5,
  "target": "000000000000000000028c610000000000000000000000000000000000000000",
  "networkhashps": 7.543343214516483e+20,
  "pooledtx": 43775,
  "chain": "main",
  "next": {
    "height": 880364,
    "bits": "17028c61",
    "difficulty": 110451907374649.5,
    "target": "000000000000000000028c610000000000000000000000000000000000000000"
  },
  "warnings": [
    "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications"
  ]
}

bitcoin (2024/12/gettarget)# build/src/bitcoin-cli getchainstates
{
  "headers": 880364,
  "chainstates": [
    {
      "blocks": 880364,
      "bestblockhash": "00000000000000000000495b9c155cdefc421a7acb06d719bc469813ccdd2193",
      "bits": "17028c61",
      "target": "000000000000000000028c610000000000000000000000000000000000000000",
      "difficulty": 110451907374649.5,
      "verificationprogress": 0.9999982707049025,
      "coins_db_cache_bytes": 8388608,
      "coins_tip_cache_bytes": 461373440,
      "validated": true
    }
  ]
}

```
</details>

Comments are suggestions for improvement!

Sjors commented:
> The addition of nbits and target to getchainstates RPC result

I didn't change that today.

tdb3 reviewed (approved):
code review re ACK a4df12323c4e9230bca58562ba17ecee4233f8fe

Makes sense to refocus away from gettarget

ryanofsky reviewed (approved):
Code review ACK a4df12323c4e9230bca58562ba17ecee4233f8fe. Only overall changes since last review were dropping new `gettarget` method and dropping changes to `getdifficulty`, but there were also various internal changes splitting and rearranging commits.

darosior commented on test/functional/data/README.md:
````diff
@@ -0,0 +1,42 @@
+```sh
+for i in {1..2015}
+do
+ faketime "`date -d @"$(( 1231006505 + $i * 120 ))"  +'%Y-%m-%d %H:%M:%S'`" \
+ bitcoind -connect=0 -nocheckpoints -stopatheight=$i
````
This documentation is incomplete, as following the instructions one would run into the check for connections performed by `getblocktemplate`:
https://github.com/bitcoin/bitcoin/blob/45719390a1434ad7377a5ed05dcd73028130cf2d/src/rpc/mining.cpp#L764-L773

To actually follow these instructions on mainnet it is necessary to comment out these lines and recompile.

Sjors replied:
Indeed I left out a few details because it would get tedious.

Sjors replied:
I ended up wasting my own time in #33446 forgetting this, so added more instructions :-)

Sjors commented on test/functional/data/README.md:
````diff
@@ -0,0 +1,42 @@
+The CPU miner is kept running as follows:
+
+```sh
+./minerd --coinbase-addr 1NQpH6Nf8QtR2HphLRcvuVqfhXBXsiWn8r --no-stratum --algo sha256d --no-longpoll --scantime 3 --retry-pause 1
````
@darosior depending on what operating system you're using, you may need one or more of these patches: https://github.com/pooler/cpuminer/pulls

Sjors commented on test/functional/data/README.md:
````diff
@@ -0,0 +1,42 @@
+pkh(xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi/44h/0h/0h/<0;1>/*)#fkjtr0yn
+```
+
+It uses `pkh()` because `tr()` outputs at low heights are not spendable (`unexpected-witness`).
````
If you do want to mine segwit or taproot outputs (which I did in the first attempt) you'll need https://github.com/pooler/cpuminer/pull/293

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -164,6 +164,7 @@ UniValue blockheaderToJSON(const CBlockIndex& tip, const CBlockIndex& blockindex
     result.pushKV("mediantime", blockindex.GetMedianTimePast());
     result.pushKV("nonce", blockindex.nNonce);
     result.pushKV("bits", strprintf("%08x", blockindex.nBits));
+    result.pushKV("target", GetTarget(tip, pow_limit).GetHex());
```
In 2a7bfebd5e788e1d9e7e07a9f1b8e3625a0301cd _Add target to getblock(header) in RPC and REST_: as pointed out in #33440 this is a bug. Unlike `getmininginfo` in baa504fdfaff4a9f61bc939035df5d5f2978cfd7 and `getblockchaininfo` in f153f57acc9f9a6f84af161d5bed9aa9965abaa3, we should not use the `tip` here but `blockindex` (like for nBits).

pracakjarenthom777-cell commented:
bf7996cbc3becf329d8b1cd2f1007fec9b3a3188
